### PR TITLE
I/O-Record Support

### PR DESCRIPTION
**Pull request by blastmaster**
These commits adds support for the OTF2 I/O-Records.
Trace writing functionality is untested and there are still some TODOs.
Trace reading seems to work fine ;)

### CODE REVIEW

File: include/otf2xx/writer/global.hpp [hunk 1]
```diff
@@ -604,7 +701,7 @@ namespace writer
         }
 
     public:
-        ~global()
+        ~global() //TODO
```
Comment (bmario) [inline include/otf2xx/writer/global.hpp]: We REALLY need to fix the missing store calls.
Definitions with one id space in a single `store(...)`

File: include/otf2xx/attribute_list.hpp [hunk 2]
```diff
@@ -526,6 +549,28 @@ namespace detail
...
     };
+
+    template<>
+    struct get_attribute<otf2::definition::attribute::attribute_type::io_file>
+    {
+        template<bool Test = false>
+        void operator()(OTF2_AttributeList* list,
+                        const otf2::definition::attribute& attribute)
+        {
+            static_assert(Test, "Not implemented yet.");
```
Comment (bmario) [inline include/otf2xx/attribute_list.hpp]: We need to fix that

File: include/otf2xx/attribute_list.hpp [hunk 3]
```diff
@@ -526,6 +549,28 @@ namespace detail
...
+    };
+
+    template<>
+    struct get_attribute<otf2::definition::attribute::attribute_type::io_handle>
+    {
+        template<bool Test = false>
+        void operator()(OTF2_AttributeList* list,
+                        const otf2::definition::attribute& attribute)
+        {
+            static_assert(Test, "Not implemented yet.");
```
Comment (bmario) [inline include/otf2xx/attribute_list.hpp]: We need to fix that

File: include/otf2xx/definition/detail/io_handle_base.hpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+namespace definition
+{
+    namespace detail
+    {
+        /**
+         * \brief Dummy class to have io handles and io handle pre created
+         * states in same id space
+         */
+        class io_handle_base
+        {
```
Comment (bmario) [inline include/otf2xx/definition/detail/io_handle_base.hpp]: If I understand the documentation right, these work like properties, i.e., each IoHandlePerCreatedState references exactly one IoHandle definition. So they don't occupy the same ID space, thus this isn't needed.

File: include/otf2xx/definition/detail/io_pre_created_handle_state_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    {
+
+        class io_pre_created_handle_state_impl : public impl_base<io_pre_created_handle_state_impl>
+        {
+        public:
+            using io_access_mode_type = otf2::common::io_access_mode_type;
+            using io_status_flag_type = otf2::common::io_status_flag_type;
+
+            io_pre_created_handle_state_impl(otf2::reference<io_handle_base> ref,
+            //io_pre_created_handle_state_impl(const otf2::definition::io_handle& handle,
```
Comment (bmario) [inline include/otf2xx/definition/detail/io_pre_created_handle_state_impl.hpp]: This should use the the commented out code.

File: include/otf2xx/definition/io_pre_created_handle_state.hpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        typedef detail::base<io_pre_created_handle_state> base;
+        typedef typename otf2::traits::definition_impl_type<io_pre_created_handle_state>::type impl_type;
+        using base::base;
+
+    public:
+        using access_mode_type = impl_type::io_access_mode_type;
+        using status_flag_type = impl_type::io_status_flag_type;
+
+        io_pre_created_handle_state(otf2::reference<detail::io_handle_base> ref,
+        //io_pre_created_handle_state(const otf2::definition::io_handle& handle,
```
Comment (bmario) [inline include/otf2xx/definition/io_pre_created_handle_state.hpp]: Should use the commented out code instead

File: include/otf2xx/definition/pre_fwd.hpp [hunk 7]
```diff
@@ -67,6 +67,14 @@ namespace definition
         class calling_context_impl;
         class interrupt_generator_impl;
 
+        class io_file_base;
+        class io_handle_base;
```
Comment (bmario) [inline include/otf2xx/definition/pre_fwd.hpp]: not needed

File: include/otf2xx/definition/pre_fwd.hpp [hunk 8]
```diff
@@ -131,6 +145,7 @@ namespace definition
     using location_group_property = property<location_group>;
     using system_tree_node_property = property<system_tree_node>;
     using calling_context_property = property<calling_context>;
+    using io_file_property = property<io_file>;
```
Comment (bmario) [inline include/otf2xx/definition/pre_fwd.hpp:147]: This looks suspicious, as io_file is polymorphic in OTF2. A property on a polymorphic definition is a new thing, we may end up with this PR to add a KNOWN_ISSUE.
  

- Edit: This is actually fine now, as io_file is also polymorphic in code.

File: include/otf2xx/writer/global.hpp [hunk 9]
```diff
@@ -399,6 +399,73 @@ namespace writer
...
+
+        void store(const otf2::definition::io_paradigm& data)
+        {
+            check(OTF2_GlobalDefWriter_WriteIoParadigm(
+                        wrt, data.ref(), data.identification().ref(),
+                        data.name().ref(),
+                        static_cast<OTF2_IoParadigmClass>(data.paradigm_class()),
+                        static_cast<OTF2_IoParadigmFlag>(data.paradigm_flags()),
+                        data.num_properties(),
+                        NULL, NULL, NULL),  //TODO: we doesn't have these arguments.
```
Comment (bmario) [inline include/otf2xx/writer/global.hpp]: We need to add these...

**Review by bmario — CHANGES_REQUESTED** [review]
Looks good overall, but I wonder if there are more changes in OTF2 2.1 we need to address prior to bumping the version to it.

File: include/otf2xx/writer/global.hpp (hunk 1 above)
Comment (blastmaster) [reply include/otf2xx/writer/global.hpp]: Right, I was not sure if the order of store calls matter?

File: include/otf2xx/attribute_list.hpp (hunk 2 above)
Comment (bmario) [reply include/otf2xx/attribute_list.hpp]: Okay, this is actually worth a separate issue. There isn't a quick fix for that.

File: include/otf2xx/writer/global.hpp (hunk 1 above)
Comment (bmario) [reply include/otf2xx/writer/global.hpp]: Yes, the sequence does matter.

**Pull request merged**